hivemq-edge/src/main/java/com/hivemq/HiveMQEdgeBootstrap.java

```diff
@@ -0,0 +1,229 @@
+package com.hivemq;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.base.Preconditions;
+import com.hivemq.api.resources.GenericAPIHolder;
+import com.hivemq.bootstrap.HiveMQExceptionHandlerBootstrap;
+import com.hivemq.bootstrap.LoggingBootstrap;
+import com.hivemq.bootstrap.ioc.DaggerInjector;
+import com.hivemq.bootstrap.ioc.Injector;
+import com.hivemq.bootstrap.ioc.Persistences;
+import com.hivemq.bootstrap.services.CompleteBootstrapService;
+import com.hivemq.bootstrap.services.CompleteBootstrapServiceImpl;
+import com.hivemq.bootstrap.services.GeneralBootstrapServiceImpl;
+import com.hivemq.bootstrap.services.PersistenceBootstrapService;
+import com.hivemq.bootstrap.services.PersistenceBootstrapServiceImpl;
+import com.hivemq.common.shutdown.ShutdownHooks;
+import com.hivemq.configuration.ConfigurationBootstrap;
+import com.hivemq.configuration.HivemqId;
+import com.hivemq.configuration.info.SystemInformation;
+import com.hivemq.configuration.service.ConfigurationService;
+import com.hivemq.edge.HiveMQCapabilityService;
+import com.hivemq.edge.impl.capability.CapabilityServiceImpl;
+import com.hivemq.edge.modules.ModuleLoader;
+import com.hivemq.exceptions.HiveMQEdgeStartupException;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+import com.hivemq.extensions.core.CommercialModuleLoaderDiscovery;
+import com.hivemq.extensions.core.HandlerService;
+import com.hivemq.extensions.core.PersistencesService;
+import com.hivemq.extensions.core.RestComponentsService;
+import com.hivemq.extensions.core.RestComponentsServiceImpl;
+import com.hivemq.metrics.MetricRegistryLogger;
+import com.hivemq.persistence.PersistenceStartup;
+import com.hivemq.persistence.connection.ConnectionPersistence;
+import com.hivemq.persistence.connection.ConnectionPersistenceImpl;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Objects;
+
+public class HiveMQEdgeBootstrap {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(HiveMQEdgeBootstrap.class);
+
+    private final @NotNull MetricRegistry metricRegistry;
+    private final @NotNull SystemInformation systemInformation;
+    private final @NotNull ModuleLoader moduleLoader;
+    private @Nullable ConfigurationService configService;
```

Comment:
I am not too familiar with the structure of invocations but looking at this field and how it is set/accessed it might be accessed from multiple threads.
Since init is delayed this field should probably be volatile.

Reply:
It should only be accessed by the main thread. Still making it volatile makes sense to be future-proof.